segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/AppendProcessor.java

```diff
@@ -203,8 +209,16 @@ public void setupAppend(SetupAppend setupAppend) {
                                     // Last event number stored according to Segment store.
                                     long eventNumber = attributes.getOrDefault(writerAttributeId, Attributes.NULL_ATTRIBUTE_VALUE);
 
-                                    String indexSegment = getIndexSegmentName(newSegment);
-                                    populateEventSizeForIndexSegmentAppends(writer, indexSegment, setupAppend.getRequestId());
+                                    // create index segment if not exist and update index segment attribute in cache
+                                    if (!isTransientSegment(newSegment) && !isTransactionSegment(newSegment)) {
+                                        Map<AttributeId, Long> eventCountAttribute = store.getAttributes(newSegment, Collections.singleton(EVENT_COUNT), true, TIMEOUT).join();
+                                        String eventCountIdentifier = (newSegment + EVENT_COUNT_SUFFIX).trim();
+                                        // storing main segments event count
+                                        long eventCount = eventCountAttribute.getOrDefault(EVENT_COUNT, 0L);
+                                        this.writerStates.put(Pair.of(eventCountIdentifier, writer), new WriterState(eventCount));
+                                        String indexSegment = getIndexSegmentName(newSegment);
+                                        createIndexSegmentIfNotExists(writer, indexSegment, setupAppend.getRequestId());
```

Comment:
This method checks to see if the segment already exists and if it does not does nothing. In this case it would be better to avoid the above `store.getAttributes` invocation, as it will exist vast majority of the time.